waspc/src/Wasp/SemanticVersion/Version.hs

```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE DeriveLift #-}
+
+module Wasp.SemanticVersion.Version
+  ( Version (..),
+    nextBreakingChangeVersion,
+    nextMajorVersion,
+    nextMinorVersion,
+    nextPatchVersion,
+    parseVersion,
+    versionParser,
+    v,
+  )
+where
+
+import qualified Language.Haskell.TH.Quote as TH
+import qualified Language.Haskell.TH.Syntax as TH
+import Numeric.Natural (Natural)
+import Text.Parsec (ParseError, Parsec, char, parse, sepBy)
+import Text.Parsec.Language (emptyDef)
+import Text.Parsec.Token (makeTokenParser, natural)
+import Text.Printf (printf)
+import Wasp.Util.TH (quasiQuoterFromParser)
+
+data Version = Version
+  { major :: !Natural,
+    minor :: !Natural,
+    patch :: !Natural
+  }
+  deriving (Eq, Ord, TH.Lift)
+
+-- | We rely on this `show` implementation to produce valid semver representation of version.
+instance Show Version where
+  show (Version mjr mnr ptc) = printf "%d.%d.%d" mjr mnr ptc
+
+parseVersion :: String -> Either ParseError Version
+parseVersion = parse versionParser ""
+
+versionParser :: Parsec String () Version
+versionParser = do
+  naturalP `sepBy` char '.' >>= \case
+    [a] -> return $ Version a 0 0
+    [a, b] -> return $ Version a b 0
+    [a, b, c] -> return $ Version a b c
+    _invalidFormat -> fail "Invalid version format"
+  where
+    naturalP = fromIntegral <$> natural lexer
+    lexer = makeTokenParser emptyDef
+
+v :: TH.QuasiQuoter
+v = quasiQuoterFromParser parseVersion
+
+nextBreakingChangeVersion :: Version -> Version
+nextBreakingChangeVersion = nextMajorVersion
+
+nextMajorVersion :: Version -> Version
+nextMajorVersion = \case
+  (Version 0 0 x) -> Version 0 0 (succ x)
+  (Version 0 x _) -> Version 0 (succ x) 0
+  (Version x _ _) -> Version (succ x) 0 0
```

Comment:
This is a bit of a misnomer. The correct name would be `nextBreakingChangeVersion` (which you have above, but is not quite the same thing).

The thing is, the semantics of what's a breaking change depend on the zeroes. The major part of the number doesn't. It's always the first one (the same applies to minor and patch).

Therefore, if we're sticking with the standard nomenclature:
```haskell
nextMajorVersion :: Version -> Version
nextMajorVersion (Version x _ _) = Version (succ x) 0 0

-- The same thing for the rest of them

nextBreakingChangeVersion :: Version -> Version
nextBreakingChangeVersion = \case
  (Version 0 0 x) -> Version 0 0 (succ x)
  (Version 0 x _) -> Version 0 (succ x) 0
  (Version x _ _) -> Version (succ x) 0 0 -- or call nextMajorVersion
```

Also, it seems to me we don't need the `nextMinorVersion`, `nextMajorVersion` and `nextPathVersion` functions. Two of them aren't used anywhere, and the only place where we use `nextMajorVersion` seems to be a better fit for `nextBreakingChangeVersion` (which we'll have to change anyway if we fix the implementation).

Therefore, I recommend removing all functions except `nextBreakingChangeVersion`.

Reply:
You are right, I went a bit too far when I was doing this I guess! This is a bit older code, but I agree now is a good moment to fix it. Done!

Reply:
Nice try you sly dog, but the old code was correct :smile:

Reply:
Oh wow you are right, I added all this in this new PR, can't believe I forgot that :D. Hm what was I thinking.